packages/adblocker/src/preprocessor.ts

```diff
@@ -0,0 +1,241 @@
+import { StaticDataView, sizeOfUTF8 } from './data-view';
+import { fastStartsWith } from './utils';
+
+export type EnvKeys =
+  | 'ext_ghostery'
+  | 'ext_abp'
+  | 'ext_ublock'
+  | 'ext_ubol'
+  | 'ext_devbuild'
+  | 'env_chromium'
+  | 'env_edge'
+  | 'env_firefox'
+  | 'env_mobile'
+  | 'env_safari'
+  | 'env_mv3'
+  | 'false'
+  | 'cap_html_filtering'
+  | 'cap_user_stylesheet'
+  | 'adguard'
+  | 'adguard_app_android'
+  | 'adguard_app_ios'
+  | 'adguard_app_mac'
+  | 'adguard_app_windows'
+  | 'adguard_ext_android_cb'
+  | 'adguard_ext_chromium'
+  | 'adguard_ext_edge'
+  | 'adguard_ext_firefox'
+  | 'adguard_ext_opera'
+  | 'adguard_ext_safari'
+  | (string & {});
+
+export class Env extends Map<EnvKeys, boolean> {}
+
+export const enum PreprocessorTypes {
+  INVALID = 0,
+  BEGIF = 1,
+  ELSE = 2,
+  ENDIF = 3,
+}
+
+export function detectPreprocessor(line: string) {
+  if (line.charCodeAt(1) !== 35 /* '#' */) {
+    return PreprocessorTypes.INVALID;
+  }
+
+  const command = line.slice(2);
+
+  if (fastStartsWith(command, 'if ')) {
```

Comment:
Nit. I think we can afford normal `startsWith` here given it's not a hot path (and probably we don't need the slicing above). I think this will make things more readable.
```suggestion
  if (line.startsWith('!#if ')) {
```